### PR TITLE
Fix discovery exception when resolving the gateway's hostname

### PR DESCRIPTION
**Pull request by cristipogacean**
fix for https://github.com/Azure/Industrial-IoT/issues/743
* discovery uses how gateway hostname environment variable
* log reports refined

### CODE REVIEW

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs [hunk 1]
```diff
@@ -512,14 +518,20 @@ public class DiscoveryServices : IDiscoveryServices, IScannerServices, IDisposab
...
+                        _logger.Information("Gateway Host Name not set");
+                        return;
+                    }
+                    _logger.Debug("Resolve IP for Gateway Host Name: {address}", hostName);
+                    var entry = await Dns.GetHostEntryAsync(hostName);
                     foreach (var address in entry.AddressList
                                 .Where(a => a.AddressFamily == AddressFamily.InterNetwork)
                                 .Where(a => !addresses.Any(b => a.Equals(b)))) {
-                        _logger.Information("Including host address {address}", address);
+                        _logger.Information("Include gateway host address {address}", address);
```
Comment (luiscantero) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs:534]: I'm wondering if "Including" would be a better word. With "Include" I'm not sure what is happening.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs [hunk 2]
```diff
@@ -512,14 +518,20 @@ public class DiscoveryServices : IDiscoveryServices, IScannerServices, IDisposab
         /// <returns></returns>
         private async Task AddLoopbackAddressesAsync(List<IPAddress> addresses) {
             // Check local host
+            string hostName = Environment.GetEnvironmentVariable(IoTEdgeVariables.IOTEDGE_GATEWAYHOSTNAME);
             try {
                 if (Host.IsContainer) {
                     // Resolve docker host since we are running in a container
-                    var entry = await Dns.GetHostEntryAsync(kDockerHostName);
+                    if (string.IsNullOrEmpty(hostName)) {
+                        _logger.Information("Gateway Host Name not set");
```
Comment (luiscantero) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs:526]: We could change to "host name" to have consistent casing

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs [hunk 3]
```diff
@@ -512,14 +518,20 @@ public class DiscoveryServices : IDiscoveryServices, IScannerServices, IDisposab
...
+                        return;
+                    }
+                    _logger.Debug("Resolve IP for Gateway Host Name: {address}", hostName);
```
Comment (luiscantero) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs:529]: "gateway host name" for consistent casing

**Review by luiscantero — APPROVED** [review]

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs (hunk 1 above)
Comment (cristipogacean) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs]: thanks! done

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs (hunk 2 above)
Comment (cristipogacean) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs]: done

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs (hunk 3 above)
Comment (cristipogacean) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs]: done

**Review by luiscantero — APPROVED** [review]

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs [hunk 4]
```diff
@@ -491,13 +492,18 @@ public class DiscoveryServices : IDiscoveryServices, IScannerServices, IDisposab
...
                             continue;
                         }
                         break;
                     }
-                    catch (Exception e) {
-                        _logger.Warning(e, "Failed to resolve the host for {discoveryUrl}", discoveryUrl);
+                    catch (SocketException se) {
+                        _logger.Warning("Failed to resolve the host for {discoveryUrl} due to {message}",
+                            discoveryUrl, se.Message);
+                        return list;
```
Comment (jschweda) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs:503]: shouldn't this catch be in the loop and we should continue to resolve other adresses?

Comment (cristipogacean) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs:503]: I find this construction rather weird... Actually we only have a single "host" string as argument, so the while is superflous. We want though to handle the special case when this argument is "localhost". "Localhost" in the module context makes no sense but we want to target the discovery call to the iotedge host system itself. I would immediately get rid of the "while" and handle "localhost" redirection before resolving the host. But, I am not a fan of on-the-fly off-context refactorings so I'll leave it be,  for now :)

**Review by MaidaB — APPROVED** [review]

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs (hunk 4 above)
Comment (jschweda) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Discovery/src/Services/DiscoveryServices.cs:503]: thanks for the clarification

**Pull request merged**